src/number_theory/arithmetic_function.lean

```diff
@@ -231,5 +242,269 @@ instance [comm_ring α] : comm_ring (arithmetic_function α) :=
+    { transitivity nat.gcd (b1 * b2) (a1 * b1),
+      {rw [mul_comm, nat.gcd_mul_right,
+           cop.coprime_mul_right.coprime_mul_left_right.symm.gcd_eq_one, one_mul]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.2.1, h.1, mul_comm c1 d1, nat.gcd_mul_left,
+            cop.coprime_mul_right.coprime_mul_left_right.symm.gcd_eq_one, mul_one] } },
+    { transitivity nat.gcd (b1 * b2) (a2 * b2),
+      {rw [nat.gcd_mul_right,
+           cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.2.1, h.2, nat.gcd_mul_right,
+            cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul] } } },
+  { rintros ⟨b1, b2⟩ h,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at h,
+    use ((b1.gcd m, b2.gcd m), (b1.gcd n, b2.gcd n)),
+    simp only [exists_prop, prod.mk.inj_iff, ne.def, mem_product, mem_divisors_antidiagonal],
+    rw [← cop.gcd_mul _, ← cop.gcd_mul _, ← h.1, nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop h.1,
+        nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop.symm _],
+    { rw [nat.mul_eq_zero, decidable.not_or_iff_and_not] at h, simp [h.2.1, h.2.2] },
+    rw [mul_comm n m, h.1] }
+end⟩
+
+end multiplicative
+
+section special_functions
+
+/-- The identity on `ℕ` as an `arithmetic_function`.  -/
+def id : arithmetic_function ℕ := ⟨id, rfl⟩
+
+@[simp]
+lemma id_apply {x : ℕ} : id x = x := rfl
+
+/-- `ζ 0 = 0`, otherwise `ζ x = 1`. The Dirichlet Series is the Riemann ζ.  -/
+def zeta [has_zero α] [has_one α] : arithmetic_function α := ⟨λ x, ite (x = 0) 0 1, rfl⟩
+
+localized "notation `ζ` := zeta" in arithmetic_function
+
+@[simp]
+lemma zeta_apply [has_zero α] [has_one α] {x : ℕ} : (ζ x : α) = if (x = 0) then 0 else 1 := rfl
+
+lemma zeta_apply_ne [has_zero α] [has_one α] {x : ℕ} (h : x ≠ 0) : ζ x = (1 : α) := if_neg h
+
+theorem zeta_mul_apply [semiring α] {f : arithmetic_function α} {x : ℕ} :
+  (ζ * f) x = ∑ i in divisors x, f i :=
+begin
+  rw mul_apply,
+  transitivity ∑ i in divisors_antidiagonal x, f i.snd,
+  { apply sum_congr rfl,
+    intros i hi,
+    rcases mem_divisors_antidiagonal.1 hi with ⟨rfl, h⟩,
+    rw [zeta_apply_ne (left_ne_zero_of_mul h), one_mul] },
+  { apply sum_bij (λ i h, prod.snd i),
+    { rintros ⟨a, b⟩ h, simp [snd_mem_divisors_of_mem_antidiagonal h] },
+    { rintros ⟨a, b⟩ h, refl },
+    { rintros ⟨a1, b1⟩ ⟨a2, b2⟩ h1 h2 h,
+      dsimp at h,
+      rw h at *,
+      rw mem_divisors_antidiagonal at *,
+      ext, swap, {refl},
+      simp only [prod.fst, prod.snd] at *,
+      apply nat.eq_of_mul_eq_mul_right _ (eq.trans h1.1 h2.1.symm),
+      rcases h1 with ⟨rfl, h⟩,
+      apply nat.pos_of_ne_zero (right_ne_zero_of_mul h) },
+    { intros a ha,
+      rcases mem_divisors.1 ha with ⟨⟨b, rfl⟩, ne0⟩,
+      use (b, a),
+      simp [ne0, mul_comm] } }
+end
+
+theorem mul_zeta_apply [semiring α] {f : arithmetic_function α} {x : ℕ} :
+  (f * ζ) x = ∑ i in divisors x, f i :=
```

Comment:
Can you deduce this from the preceding lemma by using the opposite ring of `\alpha`?

Reply:
Yes, I made that change, and it's a little shorter. Still not that short, so at some point I should either spin off the techniques from that short proof into more definitions and lemmas, or just stick to `comm_semiring`s. Not expecting much noncommutativity in this number theory.